examples/nidaqmx_analog_input/measurement.py

```diff
@@ -15,46 +24,106 @@
 @measurement_service.configuration("number_of_samples", nims.DataType.UInt64, 100)
 @measurement_service.output("acquired_samples", nims.DataType.DoubleArray1D)
-def measure(physical_channel, sample_rate, number_of_samples):
+def measure(pin_name, sample_rate, number_of_samples):
     """Perform a finite analog input measurement with NI-DAQmx."""
     logging.info(
-        "Executing measurement: physical_channel=%s sample_rate=%g number_of_samples=%d",
-        physical_channel,
+        "Executing measurement: pin_name=%s sample_rate=%g number_of_samples=%d",
+        pin_name,
         sample_rate,
         number_of_samples,
     )
-
-    def cancel_callback():
-        logging.info("Canceling measurement")
-        task_to_abort = task
-        if task_to_abort is not None:
-            task_to_abort.control(nidaqmx.constants.TaskMode.TASK_ABORT)
-
-    measurement_service.context.add_cancel_callback(cancel_callback)
-    time_remaining = measurement_service.context.time_remaining
-
-    timeout = min(time_remaining, 10.0)
-    with nidaqmx.Task() as task:
-        task.ai_channels.add_ai_voltage_chan(physical_channel)
+    session_management_client = nims.session_management.Client(
+        grpc_channel=measurement_service.get_channel(
+            provided_interface=nims.session_management.GRPC_SERVICE_INTERFACE_NAME,
+            service_class=nims.session_management.GRPC_SERVICE_CLASS,
+        )
+    )
+    with contextlib.ExitStack() as stack:
+        reservation = stack.enter_context(
+            session_management_client.reserve_sessions(
+                context=measurement_service.context.pin_map_context,
+                pin_or_relay_names=[pin_name],
+                instrument_type_id=nims.session_management.INSTRUMENT_TYPE_NI_DAQMX,
+                # If another measurement is using the session, wait for it to complete.
+                # Specify a timeout to aid in debugging missed unreserve calls.
+                # Long measurements may require a longer timeout.
+                timeout=60,
+            )
+        )
+        if len(reservation.session_info) != 1:
+            measurement_service.context.abort(
+                grpc.StatusCode.INVALID_ARGUMENT,
+                f"Unsupported number of sessions: {len(reservation.session_info)}",
+            )
+        session_info = reservation.session_info[0]
+        task = stack.enter_context(_create_nidaqmx_task(session_info))
+
+        def cancel_callback():
+            logging.info("Canceling measurement")
+            task_to_abort = task
+            if task_to_abort is not None:
+                task_to_abort.control(nidaqmx.TaskMode.TASK_ABORT)
+
+        measurement_service.context.add_cancel_callback(cancel_callback)
+        time_remaining = measurement_service.context.time_remaining
+
+        timeout = min(time_remaining, 10.0)
+        if not session_info.session_exists:
+            task.ai_channels.add_ai_voltage_chan(session_info.channel_list)
```

Comment:
Understood. Updated as suggested.

Reply:
> 1. `task: Optional[nidaqmx.Task] = None` before setting up cancel callback.

This is code to execute before setting up the cancel callback. It's not a parameter to the cancel callback. The cancel callback is a closure that accesses variables from the enclosing scope.

Reply:
Ok. Understood. As suggested, I have moved `task` before setting up `cancel_callback()` instead of having it as callback's parameter. Also, I did test with the updated changes with **NI I/O trace** and it is working as expected.